parlai/agents/bert_ranker/crossencoder_ranker.py

```diff
@@ -0,0 +1,131 @@
+                                                self.opt["type_optimization"],
+                                                total_iterations,
+                                                0.05,  # 5% scheduled warmup.
+                                                self.opt["learningrate"])
+
+    def score_candidates(self, batch, cand_vecs):
+        # concatenate text and candidates (not so easy)
+        # unpad and break
+        nb_cands = cand_vecs.size()[1]
+        size_batch = cand_vecs.size()[0]
+        text_vec = pad_left(batch.text_vec, self.NULL_IDX)
+        tokens_context = text_vec.unsqueeze(
+            1).expand(-1, nb_cands, -1).contiguous().view(nb_cands * size_batch, -1)
+        segments_context = tokens_context * 0
+
+        # remove the start token ["CLS"] from candidates
+        tokens_cands = cand_vecs.view(nb_cands * size_batch, -1)
+        segments_cands = tokens_cands * 0 + 1
+        all_tokens = torch.cat([tokens_context, tokens_cands], 1)
+        all_segments = torch.cat([segments_context, segments_cands], 1)
+        all_mask = (all_tokens != self.NULL_IDX).long()
+        all_tokens *= all_mask
+        scores = self.model(all_tokens, all_segments, all_mask)
+        return scores.view(size_batch, nb_cands)
+
+    @staticmethod
+    def dictionary_class():
+        return BertDictionaryAgent
+
+    def vectorize(self, obs, add_start=True, add_end=True, truncate=None,
+                  split_lines=False):
+        return super().vectorize(
+            obs,
+            add_start=True,
+            add_end=True,
+            truncate=self.opt["truncate"])
+
+    def _set_text_vec(self, obs, truncate, split_lines):
+        super()._set_text_vec(obs, truncate, split_lines)
+        # concatenate the [CLS] and [SEP] tokens
+        if obs is not None and "text_vec" in obs:
+            obs["text_vec"] = surround(obs["text_vec"], self.START_IDX, self.END_IDX)
+        return obs
+
+    def receive_metrics(self, metrics_dict):
+        """ Inibiting the scheduler.
+        """
+        pass
+
+    def save(self, path=None):
+        """ The state dict of DataParallel(MyModule) is not the same as
+            MyModule. In the case where we use multigpu, save it differently.
+        """
+        if not self.multigpu:
+            return super().save(self, path)
+        path = self.opt.get('model_file', None) if path is None else path
+        if path:
+            states = {}
+            states['model'] = self.model.module.state_dict()
+            states['optimizer'] = self.optimizer.state_dict()
+            with open(path, 'wb') as write:
+                torch.save(states, write)
+            # save opt file
+            with open(path + '.opt', 'w') as handle:
+                if hasattr(self, 'model_version'):
+                    self.opt['model_version'] = self.model_version()
+                json.dump(self.opt, handle)
+
+
+def pad_left(context_idx, null_idx):
```

Comment:
maybe we should have an arg in TA/TRA about whether to pad to the left or the right (so you don't have to redo this)

Reply:
thoughts @stephenroller ?

Reply:
We had an option somewheeeeeree for this, because I needed it for fairseq. 🤔

Reply:
(Per discussion with Stephen Roller offline)
There is going to be some cleaning of TA API soon, so for now we can do with this local hack.

Reply:
this so frustrating lol def keep the local hack